fix(compiler-cli): do not report unused declarations coming from an imported array

crisbeto opened the pull request:
Some apps follow a pattern where they have an array of common declarations which is imported in most standalone components, but only some of the declarations are used. Such cases will currently raise the unused imports diagnostic but can be hard to fix, because it would require either removing declarations from the common array which can break other components, or copying only the necessary declarations from the array. Since neither of these solutions is great, this commit tweaks the logic for the diagnostic so that unused imports coming from _exported_ arrays are not reported (either from the same file or another one).

crisbeto commented on packages/compiler-cli/src/ngtsc/validation/src/rules/unused_standalone_imports_rule.ts:
```diff
@@ -137,4 +138,43 @@ export class UnusedStandaloneImportsRule implements SourceFileValidatorRule {
+    if (reference.getIdentityInExpression(rawImports) !== null) {
+      return true;
+    }
+
```
Disclaimer: I'm not entirely sure if this is the best way to find which array the symbols was imported by.

alxhub replied:
I think this will catch a few things you might not expect too, such as inline nested arrays `imports: [X, Y, [Target, Z]]`.

crisbeto replied:
I think that reporting `imports: [X, Y, [Target, Z]]` is the correct behavior since it's defined inline. What I want to avoid is arrays shared between components.

alxhub commented on packages/compiler-cli/src/ngtsc/validation/src/rules/unused_standalone_imports_rule.ts:
```diff
@@ -137,4 +138,43 @@ export class UnusedStandaloneImportsRule implements SourceFileValidatorRule {
+      return true;
+    }
+
+    // Otherwise only report declarations referenced through
+    // a non-exported array within the same file.
```
Do we want to have this nuance? Clear and simple rules are preferable to heuristics in a lot of cases.

JoostK replied:
I'd be fine keeping this simple, but I would call this out in a comment to at least capture that it's been considered and decided against for simplicity.

JoostK replied:
Or do you mean the nuance of differentiating between exported arrays at all?

crisbeto committed:
fix(compiler-cli): do not report unused declarations coming from an imported array

Some apps follow a pattern where they have an array of common declarations which is imported in most standalone components, but only some of the declarations are used. Such cases will currently raise the unused imports diagnostic but can be hard to fix, because it would require either removing declarations from the common array which can break other components, or copying only the necessary declarations from the array. Since neither of these solutions is great, this commit tweaks the logic for the diagnostic so that unused imports coming from _exported_ arrays are not reported (either from the same file or another one).

crisbeto reviewed (commented):
I've addressed the feedback.
